tests/Struct/FindologicProductTest.php

```diff
@@ -366,6 +372,375 @@ public function testProduct(): void
         $this->assertEquals($properties, $findologicProduct->getProperties());
     }
 
+    public function thumbnailProvider(): array
+    {
+        return [
+            '3 thumbnails 400x400, 600x600 and 1000x100, the image of width 600 is taken' => [
+                'thumbnails' => $data['cover']['media'] = [
+                    'thumbnails' => [
+                        ['width' => 400, 'height' => 400, 'highDpi' => false],
+                        ['width' => 600, 'height' => 600, 'highDpi' => false],
+                        ['width' => 1000, 'height' => 100, 'highDpi' => false]
+                    ]
+                ]
+            ],
+            '2 thumbnails 800x800 and 2000x200, the image of width 800 is taken' => [
+                'thumbnails' => $data['cover']['media'] = [
+                    'thumbnails' => [
+                        ['width' => 800, 'height' => 800, 'highDpi' => false],
+                        ['width' => 2000, 'height' => 200, 'highDpi' => false]
+                    ]
+                ]
+            ],
+            '3 thumbnails 100x100, 200x200 and 400x400, the image directly assigned to the product is taken' => [
+                'thumbnails' => $data['cover']['media'] = [
+                    'thumbnails' => [
+                        ['width' => 100, 'height' => 100, 'highDpi' => false],
+                        ['width' => 200, 'height' => 200, 'highDpi' => false],
+                        ['width' => 400, 'height' => 400, 'highDpi' => false]
+                    ]
+                ]
+            ],
+            '0 thumbnails, the image directly assigned to the product is taken' => [
+                'thumbnails' => $data['cover']['media'] = ['thumbnails' => null]
+            ]
+        ];
+    }
+    /**
+     * @dataProvider thumbnailProvider
+     */
+    public function testCorrectThumbnailImageIsExported(array $thumbnails): void
+    {
+        $productEntity = $this->createTestProduct($thumbnails);
+        $images = $this->getImages($productEntity);
+        $customerGroupEntities = $this->getContainer()
+            ->get('customer_group.repository')
+            ->search(new Criteria(), $this->salesChannelContext->getContext())
+            ->getElements();
+
+        $findologicProductFactory = new FindologicProductFactory();
+        $findologicProduct = $findologicProductFactory->buildInstance(
+            $productEntity,
+            $this->router,
+            $this->getContainer(),
+            $this->shopkey,
+            $customerGroupEntities,
+            new XMLItem('123')
+        );
+
+        $findologicImages = $findologicProduct->getImages();
+        $this->assertEquals($images, $findologicImages);
+    }
+
+    /**
+     * @return Image[]
+     */
+    private function getImages(ProductEntity $productEntity): array
```

Comment:
Don't copy the actual implementation to ensure your code works properly. Either use static strings to compare your expected data or if this is not possible, since a random URL will be generated, try to get the data directly from the database.
In addition to that, the code in this method is way to complex for a test and would on is own already require a bunch of tests.

I know this method already existed before, but even before it was a bit on the edge and now it is just way too much.

Reply:
A test should test that your code works, by knowing before the test starts, what the outcome is. Tests should prevent bugs in the implementation and here the implementation was just copied, which may result in bugs. Imagine you have this class:

```php
class Calculator {
    public function multiply(int $number1, int $number2): int
    {
        return $number1 - $number2;
    }
}
```

You would expect `multiply` to multiply these two numbers, but in this case it substracts one from another, which is a bug. But since you are copying the implementation, you aren't even aware of this bug:

```php
class CalculatorTest extends TestCase {
    public function testAddsTwoNumbersAddsThemTogether(): void
    {
        $calculator = new Calculator();
        $number1 = 3;
        $number2 = 2;

        $expectedResult $this->getMultiply($number1, $number2);
        $actualResult = $calculator->multiply($number1, $number2);

        
        $this->assertSame($expectedResult, $actualResult);
    }

    private function getMultiply(int $number1, int $number2): int
    {
        return $number1 - $number2;
    }
}
```

Although you wouldn't copy this, because it is obvious, it is less obvious when you copy more complex code. How would you know that the implementation doesn't have a bug when you test that the code you have copied from the implementation is the same? You can't without reasonable effort.
Please rethink this and make sure not to copy the actual implementation. We want tests to avoid bugs and not to tell us everything is fine, when it isn't the case.